# Practice Midterm 2
### Psych 205 Practice Midterm 2

In [27]:
require(car)
require(stats)

In [28]:
tooth = ToothGrowth

## Exercise 1. (1 point).  
### Visualize the data. Use the scatterplot command to make a scatterplot of the data with one line for each of the delivery methods. Perform a two-way anova (with interaction) to evaluate 


In [29]:
#scatterplotMatrix(~ dose + supp, data=tooth, by.group=TRUE, id.n=0,
 #                 smooth=FALSE, col=gray(c(0,0.5,0.7)))

In [30]:
table(tooth$supp)


OJ VC 
30 30 

In [31]:
table(tooth$dose)


0.5   1   2 
 20  20  20 

In [32]:
summary(tooth$len)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   4.20   13.07   19.25   18.81   25.27   33.90 

In [33]:
twoway = lm(len ~ supp*dose, data=tooth)
summary(twoway)


Call:
lm(formula = len ~ supp * dose, data = tooth)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.2264 -2.8462  0.0504  2.2893  7.9386 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   11.550      1.581   7.304 1.09e-09 ***
suppVC        -8.255      2.236  -3.691 0.000507 ***
dose           7.811      1.195   6.534 2.03e-08 ***
suppVC:dose    3.904      1.691   2.309 0.024631 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.083 on 56 degrees of freedom
Multiple R-squared:  0.7296,	Adjusted R-squared:  0.7151 
F-statistic: 50.36 on 3 and 56 DF,  p-value: 6.521e-16


In [34]:
#All terms are significant, suggesting they are related
#the reference level appears to be Supp OJ 
#taken as a whole, VC decreases tooth length relative to OJ all else equal
# interacting with dose, VC has an estimate of 3.9 as opposed to the normal dose
#slope of 7.8 - implying again that even in the interaction VC is providing less of 
#an increase in length as dose increases
#dose is positively related to tooth length. 

# Exercise 2
Test the predictive power of three nested models 
* that correspond to a single line, 
* two lines with different intercepts and 
* two lines with different intercepts and slopes.

Type in the commands to make these models and using the summary() command comment on the results.  
You comments should address both the values of the model coefficients and of the measures of goodness of fit.
Type in your comments as "comments" in this R script below the R commands.

### Single line

In [35]:
mod1 = lm(len ~ supp, data=tooth)
summary(mod1)
#R^2 and adjR2 of .06 and .04 respectively -- this is very low
#Furher, supp isn't actually significant in this model, though it is close
# with a .06 p value. This does imply that it may not be the strongest/most 
#explanator relationship
#the reference level appears to be Supp OJ 
#VC appears to slightly decrease tooth length, tho this relationship is not significant


Call:
lm(formula = len ~ supp, data = tooth)

Residuals:
     Min       1Q   Median       3Q      Max 
-12.7633  -5.7633   0.4367   5.5867  16.9367 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   20.663      1.366  15.127   <2e-16 ***
suppVC        -3.700      1.932  -1.915   0.0604 .  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 7.482 on 58 degrees of freedom
Multiple R-squared:  0.05948,	Adjusted R-squared:  0.04327 
F-statistic: 3.668 on 1 and 58 DF,  p-value: 0.06039


### Two lines, different intercept

In [36]:
mod2 = lm(len ~ supp + dose, data=tooth)
summary(mod2)
#both R2 and adjR2 increase massivly here -- going to .7 and .69 respectively
# This indicates that adding dose means our model  explains much more of the
#variability in length of teeth
#and both are higher than the single term model, so we aren't being punished so much
#for additional parameters that this isn't a benefit - it is
#Both dose and supp are significant, dose extremely significant with a much higher 
#coefficient. This seems to imply that dose was a very important term for doing 
#accurate predictions
#the reference level appears to be Supp OJ 
#VC decreases tooth length relative to OJ all else equal
# increased dose increases tooth length


Call:
lm(formula = len ~ supp + dose, data = tooth)

Residuals:
   Min     1Q Median     3Q    Max 
-6.600 -3.700  0.373  2.116  8.800 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   9.2725     1.2824   7.231 1.31e-09 ***
suppVC       -3.7000     1.0936  -3.383   0.0013 ** 
dose          9.7636     0.8768  11.135 6.31e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.236 on 57 degrees of freedom
Multiple R-squared:  0.7038,	Adjusted R-squared:  0.6934 
F-statistic: 67.72 on 2 and 57 DF,  p-value: 8.716e-16


### Two lines, different slopes

In [37]:
mod3 = lm(len ~ supp*dose, data=tooth)
summary(mod3)
#R2 and adjR2 increased - despite additional terms, it is still higher than the
#model, meaning more variability is explained even when adjusted for the additional 
#terms. All terms are significant, suggesting they are related
#the reference level appears to be Supp OJ 
#taken as a whole, VC decreases tooth length relative to OJ all else equal
# interacting with dose, VC has an estimate of 3.9 as opposed to the normal dose
#slope of 7.8 - implying again that even in the interaction VC is providing less of 
#an increase in length as dose increases
#dose is positively related to tooth length. (same description as in previous q)

#Comparing adjusted R^2, it increases from model1 -> model2 -> model3. 
#So, it appears that even with the additional parameters, adding these terms is still
#making this regression more accurate



Call:
lm(formula = len ~ supp * dose, data = tooth)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.2264 -2.8462  0.0504  2.2893  7.9386 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   11.550      1.581   7.304 1.09e-09 ***
suppVC        -8.255      2.236  -3.691 0.000507 ***
dose           7.811      1.195   6.534 2.03e-08 ***
suppVC:dose    3.904      1.691   2.309 0.024631 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.083 on 56 degrees of freedom
Multiple R-squared:  0.7296,	Adjusted R-squared:  0.7151 
F-statistic: 50.36 on 3 and 56 DF,  p-value: 6.521e-16


# Exercise 3.
How many parameters are there in each of the three models?
Make a line plot of model parameters versus R2adj. You can get R2adj from the summary commands.

In [40]:
m1len = length(coef(mod1))
#2
m2len = length(coef(mod2))
#3
m3len = length(coef(mod3))
# get adjR2
m1r2 = summary(mod1)$adj.r.squared
m2r2 = summary(mod2)$adj.r.squared
m3r2 = summary(mod3)$adj.r.squared
#plot
#plot(c(m1len,m2len,m3len),c(m1r2,m2r2,m3r2), type="l")

# Exercise 4
Calculate the R2, R2 adj, F and p value for the full model "by hand".
Make sure that the results match those you obtained from the summary command.

## GOALS
Multiple R-squared:  0.7296,	Adjusted R-squared:  0.7151 
F-statistic: 50.36 on 3 and 56 DF,  p-value: 6.521e-16

## R2

In [41]:
mod0 = lm(len ~ 1, data=tooth)
tooth_hat_full = predict(mod3)
tooth_hat_null =  predict(mod0)

ss3 = sum((tooth$len - tooth_hat_full)^2)
ss0 = sum((tooth$len - tooth_hat_null)^2)

R2 = 1 - (ss3/ss0)
print(R2)

[1] 0.7295544


## Adjusted R2

In [42]:
n = length(mod3$model$len)
k = length(mod3$coefficients)

R2adj = 1 - (ss3/(n-k))/(ss0/(n-1))
print(R2adj)

[1] 0.7150662


## F value

In [43]:
F_val = ((ss0 - ss3)/(k-1))/((ss3)/(n-k))
print(F_val)

[1] 50.35522


## P value

In [44]:
require(stats)
pfval = 1 - pf(F_val, df1 = (k-1), df2 = n-k)

# Exercise 5
Using 10 fold cross-validation estimate the cross-validated
R2 for the three models.  Make a plot of average R2CV versus number of parameters.

In [45]:
df = tooth

#find the number of folds
n_folds = 10

#get rid of nas, shuffle data
df <-na.omit(df[sample(nrow(df)),])

#create folds
folds <- cut(seq(1,nrow(df)),breaks=n_folds,labels=FALSE)

#create arrays for results
R2cv_mod1 = array(data=0, dim=n_folds)
R2cv_mod2 = array(data=0, dim=n_folds)
R2cv_mod3 = array(data=0, dim=n_folds)

#for loop
for(i in 1:n_folds){
  
  #split up data
  test = df[folds==i,]
  train = df[folds !=i,]
  
  #fit models
  mod0 = lm(len ~ 1, data=train)
  mod1 = lm(len ~ supp, data=train)
  mod2 = lm(len ~ supp + dose, data=train)
  mod3 = lm(len ~ supp*dose, data=train)
  
  yhat_mod0 = predict(mod0, data=test)
  yhat_mod1 = predict(mod1, data=test)
  yhat_mod2 = predict(mod2, data=test)
  yhat_mod3 = predict(mod3, data=test)
  
  ss0 = sum((yhat_mod0 - test$len)^2)
  ss1 = sum((yhat_mod1 - test$len)^2)
  ss2 = sum((yhat_mod2 - test$len)^2)
  ss3 = sum((yhat_mod3 - test$len)^2)
  
  R2cv_mod1[i] = abs(1-ss1/ss0)
  R2cv_mod2[i] = abs(1-ss2/ss0)
  R2cv_mod3[i] = abs(1-ss3/ss0)
  
}

meanR2_1 = mean(R2cv_mod1)
meanR2_2 = mean(R2cv_mod2)
meanR2_3 = mean(R2cv_mod3)

means = c(meanR2_1, meanR2_2, meanR2_3)
params = c(m1len, m2len, m3len)

#plot(params, means, type='l')

# Exercise 6.
First, use the appropriate classical commands to statiscally compare the single line model to
 the full model. 
 
 Second, using the results of the cross-validation in 5, determined whether you would reach the same conclusion. 
 
 For example compare the mean R2cv plus or minus two standard errors to assess whether one is significantly
 better.

### Traditional methods

In [62]:
anova(mod3, mod1)

Res.Df RSS       Df Sum of Sq F        Pr(>F)      
1 50      859.0039 NA        NA       NA           NA
2 52     2875.4312 -2 -2016.427 58.68504 7.626124e-14

In [63]:
SSfull = sum((mod3$residuals)^2)
SSnull = sum((mod1$residuals)^2)

kfull = length(mod3$coefficients)
ksimple = length(mod1$coefficients)

n = length(mod3$model$len)

#calculating difference in quality using R2 and R2adjusted
R2 = 1-SSfull/SSnull #atypical way to do this
print(R2)

R2adj = 1-(SSfull/(n-kfull))/(SSnull/(n-ksimple))
print(R2adj)


#diff method 2
R2change = summary(mod3)$r.squared - summary(mod1)$r.squared
print(R2change)

#F val and p val

q = kfull - ksimple

Fval = ((SSnull - SSfull)/q)/(SSfull/(n-kfull))

pval = pf(Fval, df1=q, df2 = n-k, lower.tail=FALSE)


[1] 0.7012608
[1] 0.6893113
[1] 0.6558694


In [64]:
Fval

[1] 58.68504

In [65]:
pval

[1] 7.626124e-14

According to these results, model 3 does work significantly better than model 1. The F value of 58 with a p-value of <.00001 implies that model 3 is significantly better than model 1, as that F value would be incredibly unlikely under the null hypothesis tha the model peformance is identical

### Non-traditional methods

In [56]:
summary(R2cv_mod1)

    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
0.003041 0.038957 0.052216 0.064746 0.077868 0.185447 

In [58]:
summary(R2cv_mod3)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.3010  0.4833  0.7414  0.7150  0.9192  1.2255 

The minimum R2cv found for model 3 is higher than the max found for model 1 - it seems quite clear that this model is significantly better. Checking using standard error, this is confirmed as well - even in the worst case scenario for model 3 described in the question (mod3 mean - 2 standard errors vs mod 1 mean + 2 standard errors), model 3 is still better. 

In [59]:
se_mod1 = sd(R2cv_mod1)/sqrt((nrow(tooth)-nrow(tooth)/10))

In [60]:
se_mod3 = sd(R2cv_mod3)/sqrt((nrow(tooth)-nrow(tooth)/10))

In [61]:
(meanR2_3 - se_mod3*2)>(meanR2_1 - se_mod1*2)

[1] TRUE